# Tarea 1 

Alumnos:

* Miguel Espinoza
* Sebastián López
* Pedro Pérez

## P1

### Parte 1

In [1]:
import os

folder = 'data/raw/'  #Carpeta principal en la que estan los archivos
weeks = os.listdir(folder)  #Se agregan las subcarpetas a una lista
weeks

['w13', 'w14', 'w15', 'w16', 'w17']

In [2]:
#Se crean dos listas para guardar los archivos normales
#y los 'furnished'
files = []
files_furnished = []

#Mediante un bucle for se guardan las direcciones de todos
#los archivos en las listas creadas
for week in weeks:
    direccion = folder + week + '/'
    file = os.listdir(direccion)
    files.append(direccion + file[0])
    files_furnished.append(direccion + file[1])
    
files

['data/raw/w13/metrocuadrado_all_w13.csv',
 'data/raw/w14/metrocuadrado_all_w14.csv',
 'data/raw/w15/metrocuadrado_all_w15.csv',
 'data/raw/w16/metrocuadrado_all_w16.csv',
 'data/raw/w17/metrocuadrado_all_w17.csv']

In [3]:
import pandas as pd

#Se lee el primer dataframe para observar los datos
data_all = pd.read_csv(files[0])
data_all.head(3)

property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6  
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2

En primer lugar se combinan los dataframes normales

In [4]:
data_all = pd.read_csv(files[0])

for file in files:
    dato_semanal = pd.read_csv(file)
    data_all = data_all.merge(dato_semanal, how = 'outer') 
    #Se pone 'outer' para hacer la union de todos los dataframes

print('Cantidad de datos: ', len(data_all))
data_all.head(3)

Cantidad de datos:  16470


property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6  
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2

In [5]:
#Se ve si hay duplicados
print('Duplicados: ', data_all.duplicated().sum())

Duplicados:  175


In [6]:
#Se eliminan las filas duplicadas
data_all.drop_duplicates(inplace=True)
print('Cantidad de datos: ', len(data_all))

Cantidad de datos:  16295


Ahora se combinan los archivos furnished

In [7]:
data_furnished = pd.read_csv(files_furnished[0])

for file in files_furnished:
    dato_semanal = pd.read_csv(file)
    data_furnished = data_furnished.merge(dato_semanal, how='outer')

print('Cantidad de datos: ', len(data_furnished))
data_furnished.head(3)

Cantidad de datos:  2101


property_type|rent_type|location       price  n_rooms  \
0  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000      3.0   
1  Casa en Venta Y Arriendo, BELLAVISTA Bogotá D.C..  $1.300.000      3.0   
2  Casa en Arriendo, PORTALES DEL NORTE Bogotá D.C..  $1.800.000      3.0   

   n_bath   surface                                            details  \
0     2.0  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
1     2.0   72.0 m2  hermosa casa en venta y en arriendo en, mandal...   
2     3.0  76.81 m2  Estrato 4. CASA TRIPLEX!! Exterior . Cuenta co...   

                                                 url  metrocuadrado_index  
0  https://www.metrocuadrado.com/inmueble/arriend...                 57.3  
1  https://www.metrocuadrado.com/inmueble/venta-c...                 41.4  
2  https://www.metrocuadrado.com/inmueble/arriend...                 41.3

In [8]:
#Se buscan los duplicados y se eliminan
print('Duplicados: ', data_furnished.duplicated().sum())

Duplicados:  2


In [9]:
data_furnished.drop_duplicates(inplace=True)

print('Cantidad de datos: ', len(data_furnished))
#Se observa que son muchos menos los datos de tipo furnished

Cantidad de datos:  2099


Al intentar combinar ambos dataframes usando `data_all.merge(data_furnished, how='outer')` aparece el error:

`ValueError: You are trying to merge on object and float64 columns. If you wish to proceed you should use pd.concat`

Por lo que se decide comprobar los tipos de las columnas

In [10]:
#Se observan las columnas de los dataframe
print('Tipos Data All')
print(data_all.dtypes)
print()
print('Tipos Data Furnished')
print(data_furnished.dtypes)

Tipos Data All
property_type|rent_type|location     object
price                                object
n_rooms                              object
n_bath                               object
surface                              object
details                              object
url                                  object
metrocuadrado_index                 float64
dtype: object

Tipos Data Furnished
property_type|rent_type|location     object
price                                object
n_rooms                             float64
n_bath                              float64
surface                              object
details                              object
url                                  object
metrocuadrado_index                 float64
dtype: object


Se observa que el error se debe a que en `Data All` las columnas `n_rooms` y `n_bath` son de tipo `object`, en cambio en `Data Furnished` son de tipo `float64`.

Este problema se resuelve simplemente cambiando el tipo de esas columnas.

In [11]:
#Se cambian las columnas de 'data_furnished' a object
data_furnished = data_furnished.astype({'n_rooms':'object', 'n_bath':'object'})

#Obs: No se pueden cambiar las columnas de 'data_all' a float64, pues hay un dato llamado '5+'

In [153]:
#Se combinan los dataframes
data = data_all.merge(data_furnished, how='outer', indicator='Furnished')

#Se coloca utiliza el parámetro indicator para indicar de que dataframe 
#proviene cada fila.
#left_only indica que viene del dataframe 'data_all'
#right_only indica que viene del dataframe 'data_furnished'
#y both que proviene de ambos

print('Cantidad de datos: ', len(data))
data.head(3)

Cantidad de datos:  18394


property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

   Furnished  
0  left_only  
1  left_only  
2  left_only

In [154]:
data['Furnished'].replace({'left_only':'No', 'right_only':'Si', 'both':'Ambos'}, inplace=True)

#Se cambian los nombres para indicar más claramente si provienen de furnished o no
data.head(3)

property_type|rent_type|location       price n_rooms  \
0    Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..  $1.050.000       3   
1  Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...  $3.930.000       3   
2  Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...    $750.000       3   

  n_bath   surface                                            details  \
0      2   63.0 m2  Excelente casa cerca de la calle 80 y Avenida ...   
1      2  100.0 m2  Se arrienda casa bonita totalmente amoblada co...   
2      1   90.0 m2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished  
0        No  
1        No  
2        No

In [155]:
#Se cuentan la cantidad de cada valores de cada tipo
data['Furnished'].value_counts()

No    16295
Si     2099
Name: Furnished, dtype: int64

Como se observa no hay elementos comunes entre ambos grupos, son disjuntos. No hay archivos con texto 'furnished' que no estén contenidos en archivos con texto 'all'.

In [156]:
#Esto se comprueba tambien haciendo una intersección entre los dataframes
#mediante el argumento 'inner'
pd.merge(data_all, data_furnished, how='inner')

#Se observa que el resultado es un dataframe vacío

Empty DataFrame
Columns: [property_type|rent_type|location, price, n_rooms, n_bath, surface, details, url, metrocuadrado_index]
Index: []

### Parte 2

#### A

Primero se limpia el precio

In [157]:
data['price'].head(3)

0    $1.050.000
1    $3.930.000
2      $750.000
Name: price, dtype: object

Se observa que son de tipo `object`, pues el precio contiene el signo peso '$' y los puntos. Convendría eliminar este signo y los puntos para poder pasar los datos a `float64`.

In [158]:
data['price'] = data['price'].str.replace('$','')
data['price'] = data['price'].str.replace('.','')
data['price'].head(3)

0    1050000
1    3930000
2     750000
Name: price, dtype: object

In [159]:
#Se cambia el tipo a float64
data['price'] = data['price'].astype('float')
data['price'].head(3)

0    1050000.0
1    3930000.0
2     750000.0
Name: price, dtype: float64

Ahora se limpia el área

In [160]:
data['surface'].tail(4)

18390    164.34 m2
18391     110.0 m2
18392     106.0 m2
18393     135.0 m2
Name: surface, dtype: object

Se deben eliminar los carácteres ' m2' al final del número, y luego cambiar el tipo a `float`.

In [161]:
data['surface'] = data['surface'].str.replace(' m2','')
data['surface'] = data['surface'].astype('float')
data['surface'].tail(4)

18390    164.34
18391    110.00
18392    106.00
18393    135.00
Name: surface, dtype: float64

Ahora se limpia el número de habitaciones

In [162]:
data['n_rooms'].head(3)

0    3
1    3
2    3
Name: n_rooms, dtype: object

Parece que fueran números enteros, pero el tipo de la columna es object

In [163]:
#Se ven los valores de la columna
data['n_rooms'].unique()

array(['3', '5', '4', '2', '1', nan, '5+', 3.0, 5.0, 4.0, 2.0, 1.0],
      dtype=object)

Se observa que la columna tiene valores de tipo string y float. Más aún se ve que se podría cambiar inmediatamente el tipo a `float` sino existiera el valor `5+`.

In [164]:
#Se cuenta cuantas entradas hay de cada valor
data['n_rooms'].value_counts()

2      6571
1      6271
3      1351
4      1198
1.0    1040
2.0     938
5       862
3.0      52
5.0      36
4.0      32
5+        9
Name: n_rooms, dtype: int64

Se ve que solo son 9 las entradas con valor `5+`, por lo que es una parte infima del total del dataset que es de $18394$ entradas, simplemente se podrían eliminar. Pero como no se desea botar datos, se cambiará el valor `5+` por $6$, dado que es el entero mayor a 5 más cercano.

In [165]:
data['n_rooms'] = data['n_rooms'].replace('5+',6)
#Se reemplazan el valor que daba problemas, y se muestran los valores de las entradas
data['n_rooms'].unique()

array(['3', '5', '4', '2', '1', nan, 6, 3.0, 5.0, 4.0, 2.0, 1.0],
      dtype=object)

In [166]:
#Como esta correcto, se puede cambiar el tipo de la columna
data['n_rooms'] = data['n_rooms'].astype('float')
data['n_rooms'].head(3)

0    3.0
1    3.0
2    3.0
Name: n_rooms, dtype: float64

Por último se limpia el número de baños

In [167]:
data['n_bath'].tail(3)

18391    3
18392    3
18393    2
Name: n_bath, dtype: object

Nuevamente parecen ser enteros, pero el tipo es object

In [168]:
#Se ven los valores de la columna
data['n_bath'].unique()

array(['2', '1', '4', '3', nan, '5', '5+', 2.0, 3.0, 4.0, 5.0, 1.0],
      dtype=object)

Nuevamente el error viene dado por `5+`, por lo que se procede de igual manera que en `n_rooms`.

In [169]:
data['n_bath'] = data['n_bath'].replace('5+',6) #Se reemplazan '5+' por 6
data['n_bath'] = data['n_bath'].astype('float') #Se cambia el tipo a float
data['n_bath'].head(3)

0    2.0
1    2.0
2    1.0
Name: n_bath, dtype: float64

Finalmente se muestran nuevamente los tipos de las columnas

In [170]:
data.dtypes

property_type|rent_type|location     object
price                               float64
n_rooms                             float64
n_bath                              float64
surface                             float64
details                              object
url                                  object
metrocuadrado_index                 float64
Furnished                            object
dtype: object

#### B

Se debe dividir la columna `property_type|rent_type|location` en 3 columnas distintas.

In [171]:
#Se hace una copia de la data por si se realiza un error
copia_seguridad = data.copy()

In [151]:
#Cuando algo salga mal correr esta celda
data = copia_seguridad.copy()

In [172]:
#Se muestran unos pocos datos de esta columna
data['property_type|rent_type|location']

0          Casa en Arriendo, LA SOLEDAD NORTE Bogotá D.C..
1        Casa en Arriendo, CIUDAD SALITRE SUR-ORIENTAL ...
2        Casa en Arriendo, VILLA DE LOS ALPES II SECTOR...
3              Casa en Arriendo, PARIS GAITAN Bogotá D.C..
4        Casa en Arriendo, BALCONES DE ORIENTE Bogotá D...
                               ...                        
18389    Apartamento en Arriendo, CHICO NORTE Bogotá D.C..
18390    Apartamento en Arriendo, CHICO NORTE II Bogotá...
18391    Apartamento en Venta Y Arriendo, CHICO NORTE B...
18392    Apartamento en Arriendo, MOLINOS NORTE Bogotá ...
18393       Apartamento en Arriendo, EL NOGAL Bogotá D.C..
Name: property_type|rent_type|location, Length: 18394, dtype: object

Se pueden deducir varias cosas:
1. El tipo de propiedad es la primera palabra del string.
2. El tipo de renta aparece después de la palabra 'en' y un espacio ' '.
3. El lugar esta después de la coma.

En primer lugar se decide dividir el string en dos partes separadas por la coma.

In [173]:
new_columns = data['property_type|rent_type|location'].str.split(pat=',', expand=True)
new_columns.head(3)

0                                           1
0  Casa en Arriendo               LA SOLEDAD NORTE Bogotá D.C..
1  Casa en Arriendo    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..
2  Casa en Arriendo   VILLA DE LOS ALPES II SECTOR Bogotá D.C..

In [174]:
#Se insertan las columnas en la data 
data[['property_type|rent_type', 'location']] = new_columns

#Se elimina la columna original
data.drop('property_type|rent_type|location', axis=1, inplace=True)

data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished property_type|rent_type  \
0        No        Casa en Arriendo   
1        No        Casa en Arriendo   
2        No        Casa en Arriendo   

                                     location  
0               LA SOLEDAD NORTE Bogotá D.C..  
1    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..  
2   VILLA DE LOS ALPES II SECTOR Bogotá D.C..

Ahora se divide la columna `property_type|rent_type` en dos columnas más.

In [175]:
new_columns = data['property_type|rent_type'].str.split(pat='en ', expand=True)
new_columns.head(3)

0         1
0  Casa   Arriendo
1  Casa   Arriendo
2  Casa   Arriendo

In [176]:
#Se insertan las columnas en la data y se elimina la columna original
data[['property','rent']] = new_columns
data.drop('property_type|rent_type', axis=1, inplace=True)
data.head(3)

price  n_rooms  n_bath  surface  \
0  1050000.0      3.0     2.0     63.0   
1  3930000.0      3.0     2.0    100.0   
2   750000.0      3.0     1.0     90.0   

                                             details  \
0  Excelente casa cerca de la calle 80 y Avenida ...   
1  Se arrienda casa bonita totalmente amoblada co...   
2  CASA DE TRES NIVELES TRES ALCOBAS SALA COMEDOR...   

                                                 url  metrocuadrado_index  \
0  https://www.metrocuadrado.com/inmueble/arriend...                 25.6   
1  https://www.metrocuadrado.com/inmueble/arriend...                 57.3   
2  https://www.metrocuadrado.com/inmueble/arriend...                 12.2   

  Furnished                                    location property      rent  
0        No               LA SOLEDAD NORTE Bogotá D.C..    Casa   Arriendo  
1        No    CIUDAD SALITRE SUR-ORIENTAL Bogotá D.C..    Casa   Arriendo  
2        No   VILLA DE LOS ALPES II SECTOR Bogotá D.C..    Casa   Arriendo

Por último se revisa que no hayan espacios vacíos al inicio o final de las nuevas columnas.

In [178]:
print(data['location'][0])
print(data['property'][0])
print(data['rent'][0])

 LA SOLEDAD NORTE Bogotá D.C..
Casa 
Arriendo


Tanto location como property tienen espacios, por lo que se utiliza strip para eliminarlos.

In [182]:
data['location'] = data['location'].str.strip()
data['property'] = data['property'].str.strip()

### Parte 3

### Parte 4

### Parte 5

### Parte 6

## P2

# P3 

## Implementación

### Parte 1

In [120]:
from sklearn.base import BaseEstimator as BE
from sklearn.base import RegressorMixin as RM
import numpy as np

class RegresionBayesianaEmpirica(BE,RM):
    
    def __init__(self,alpha,beta,tol=1e-15,maxiter=100):
        self.set_params(alpha=alpha,beta=beta,tol=tol,maxiter=maxiter)
        
    def get_posteriori(self,X,y,alpha,beta):
        N,d = X.shape
        try:
            S_N = np.linalg.inv(alpha*np.eye(d) + beta*np.dot(X.T,X)) # dim dxd
        except LinAlgError:
            print('Matriz S_N no invertible')
        m_N = beta*np.dot(S_N,np.dot(X.T,y)) # dim dx1

        return m_N, S_N
    
    
    def fit(self,X,y,i=0,verbose=True):
        i += 1
        if verbose:
            print('Iteracion',i)
            print('Valores actuales: alpha=',self.alpha,' beta=',self.beta)
        if i <= self.maxiter:
            N,d = X.shape
            eigen_val, eigen_vec = np.linalg.eig(self.beta*np.dot(X.T,X))
            eigen_val = np.real_if_close(eigen_val,tol=1)
            gamma = np.sum(eigen_val/(self.alpha + eigen_val))
            m_N, S_N = self.get_posteriori(X,y,self.alpha,self.beta)
        
            new_alpha = gamma/(np.dot(m_N.T,m_N))
            new_beta = (N-gamma)/(np.sum((y-np.dot(m_N.T,X.T))**2))
        
            if np.abs(self.alpha - new_alpha) < self.tol and np.abs(self.beta - new_beta) < self.tol:
                self.alpha = new_alpha
                self.beta = new_beta
                self.m_N = m_N
                self.S_N = S_N
                if verbose:
                    print('Valores finales: alpha=',self.alpha,' beta=',self.beta)
            else:
                self.alpha = new_alpha
                self.beta = new_beta
                self.m_N = m_N
                self.S_N = S_N
                self.fit(X,y,i)
                
    
    def predict(self,X_,return_std=False):
        y_ = np.dot(X_,self.m_N)
        y_std = np.sqrt((1/self.beta) + np.diag(np.dot(X_,np.dot(self.S_N,X_.T))))
        
        if return_std:
            return (y_,y_std)
        return y_

### Parte 2

In [164]:
import pandas as pd
import numpy as np

#Datos para prueba
df = pd.read_csv('data.csv')

In [165]:
df.replace([np.inf, -np.inf], np.nan, inplace=True)
df.dropna(inplace=True)
df.drop(df[df.surface==0].index, inplace=True)
df.drop(['price','surface','url','details','CodigoUPZ','personas','property'],axis=1,inplace=True)#

In [166]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16091 entries, 3 to 18393
Data columns (total 18 columns):
n_rooms                                    16091 non-null float64
n_bath                                     16091 non-null float64
metrocuadrado_index                        16091 non-null float64
Furnished                                  16091 non-null object
location                                   16091 non-null object
rent                                       16091 non-null object
Precio m2                                  16091 non-null float64
Garajes                                    16091 non-null int64
tipo de producto                           16091 non-null float64
trabajoinf_ninos_5_17_anos_perc            16091 non-null float64
trabajoinfampliado_ninos_5_17_anos_perc    16091 non-null float64
jovenes_14_24_anos_nini_perc               16091 non-null float64
indice_envegecimiento                      16091 non-null float64
jefe_mujer_perc                       

In [167]:
df_y = df['Precio m2']#.to_numpy()
df_X = df.drop('Precio m2',axis=1)#.to_numpy()

#num_cols = df_X.select_dtypes(include=['int64', 'float64']).columns
#cat_cols = df_X.select_dtypes(include=['object']).columns
float_cols = ['pob_density']
percentage_cols = ['n_rooms','n_bath','metrocuadrado_index','Garajes','trabajoinf_ninos_5_17_anos_perc',
              'trabajoinfampliado_ninos_5_17_anos_perc','jovenes_14_24_anos_nini_perc','indice_envegecimiento',
              'jefe_mujer_perc','adultos_mayores_pobres_perc','indice_inseguridad','areas_verdes_perc']
cat_cols = ['Furnished','location','rent','tipo de producto']

In [168]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder, PolynomialFeatures, FunctionTransformer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

#cat pipe
cat_pipe = Pipeline(steps=[('ordinal', OneHotEncoder(handle_unknown='ignore'))])

#num pipe
#num_pipe = Pipeline(steps=[('poly',PolynomialFeatures(degree=3)),('scaler',StandardScaler())])
float_pipe = Pipeline(steps=[('scaler',StandardScaler()),('poly',PolynomialFeatures(degree=3))])
percentage_pipe = Pipeline(steps=[('scaler',MinMaxScaler()),('poly',PolynomialFeatures(degree=3))])

#Se diferencian columnas numéricas y categóricas
prep = ColumnTransformer(transformers=[('flt',float_pipe,float_cols),('perc',percentage_pipe,percentage_cols),('cat',cat_pipe,cat_cols)])

### Parte 3

In [169]:
random_state = 7
def run(df_X,df_y):
    from sklearn.model_selection import train_test_split
    import numpy as np
    import pickle
    
    alpha_grid = np.linspace(1e-10,1e-5,3,endpoint=True)
    beta_grid = np.linspace(1e-10,1e-5,3,endpoint=True)
    
    actual_R2 = 0
    actual_RMSE = 0
    
    X_train, X_test, y_train, y_test = train_test_split(df_X, df_y, test_size=0.2, random_state=random_state)
    def run_single(pipeline):
        pipeline.fit(X_train,y_train)
        print('Fit terminado', '\n')
        y_pred = pipeline.predict(X_test) 

        R_2 = pipeline.score(X_test, y_test)
        RMSE = np.sqrt(np.mean((y_pred-y_test)**2))
        return R_2, RMSE
    
    best_alpha = alpha_grid[0]
    best_beta = beta_grid[0]
    for alpha in alpha_grid:
        for beta in beta_grid:
            Tubo = Pipeline(steps=[('prep',prep),('clf', RegresionBayesianaEmpirica(alpha,beta))])
            print("Empezando fit: alpha y beta iniciales",[alpha,beta])
            R_2, RMSE = run_single(Tubo)
            if R_2 > actual_R2:
                actual_R2 = R_2
                actual_RMSE = RMSE
                best_alpha = alpha
                best_beta = beta
                
    best_fit = Pipeline(steps=[('prep',prep),('clf', RegresionBayesianaEmpirica(best_alpha,best_beta))])
    best_fit.fit(X_train,y_train,clf__verbose=False)
    filename = 'modelo.pkl'
    pickle.dump(best_fit,open(filename,'wb'))
    
    print('RMSE:',actual_RMSE,' R_2:',actual_R2)

In [170]:
run(df_X,df_y)

Empezando fit: alpha y beta iniciales [1e-10, 1e-10]
Iteracion 1
Valores actuales: alpha= 1e-10  beta= 1e-10
Iteracion 2
Valores actuales: alpha= 7.426618085220108e-12  beta= 3.940169346525641e-13
Iteracion 3
Valores actuales: alpha= 3.196548133399462e-11  beta= 3.994279635346027e-13
Iteracion 4
Valores actuales: alpha= 7.720029719024391e-11  beta= 3.999414575851797e-13
Iteracion 5
Valores actuales: alpha= 1.3773188621953584e-10  beta= 3.9977698036263363e-13
Iteracion 6
Valores actuales: alpha= 1.9931039409778445e-10  beta= 3.995646776464928e-13
Iteracion 7
Valores actuales: alpha= 2.5283577006326216e-10  beta= 3.993910429944042e-13
Iteracion 8
Valores actuales: alpha= 2.958996838152452e-10  beta= 3.992651845728542e-13
Iteracion 9
Valores actuales: alpha= 3.291581172890137e-10  beta= 3.9917692032169043e-13
Iteracion 10
Valores actuales: alpha= 3.5423057282882533e-10  beta= 3.991152819944527e-13
Iteracion 11
Valores actuales: alpha= 3.728428447163209e-10  beta= 3.990720798253454e-13
Ite

Iteracion 8
Valores actuales: alpha= 1.929836710738247e-10  beta= 3.995862129568203e-13
Iteracion 9
Valores actuales: alpha= 2.475756464281259e-10  beta= 3.9940730997723335e-13
Iteracion 10
Valores actuales: alpha= 2.917635691504164e-10  beta= 3.9927670073803873e-13
Iteracion 11
Valores actuales: alpha= 3.260057702742306e-10  beta= 3.991849616248775e-13
Iteracion 12
Valores actuales: alpha= 3.5187380639633163e-10  beta= 3.991209043608223e-13
Iteracion 13
Valores actuales: alpha= 3.711029330366756e-10  beta= 3.990760299336701e-13
Iteracion 14
Valores actuales: alpha= 3.852444508575908e-10  beta= 3.9904443649815546e-13
Iteracion 15
Valores actuales: alpha= 3.955683639305274e-10  beta= 3.9902209139905854e-13
Iteracion 16
Valores actuales: alpha= 4.030668623518364e-10  beta= 3.99006229079721e-13
Iteracion 17
Valores actuales: alpha= 4.0849367343463073e-10  beta= 3.989949370658417e-13
Iteracion 18
Valores actuales: alpha= 4.124111957620663e-10  beta= 3.9898688172443793e-13
Iteracion 19
Valo

Iteracion 12
Valores actuales: alpha= 3.865180389931024e-10  beta= 3.9904164766072225e-13
Iteracion 13
Valores actuales: alpha= 3.9649511871798213e-10  beta= 3.990201144079187e-13
Iteracion 14
Valores actuales: alpha= 4.037384507336889e-10  beta= 3.9900482317671846e-13
Iteracion 15
Valores actuales: alpha= 4.0897893116710946e-10  beta= 3.9899393491663614e-13
Iteracion 16
Valores actuales: alpha= 4.127610936030931e-10  beta= 3.9898616613190016e-13
Iteracion 17
Valores actuales: alpha= 4.154860093301691e-10  beta= 3.989806148707917e-13
Iteracion 18
Valores actuales: alpha= 4.1744678050694676e-10  beta= 3.9897664388191917e-13
Iteracion 19
Valores actuales: alpha= 4.188564449763018e-10  beta= 3.989738010913319e-13
Iteracion 20
Valores actuales: alpha= 4.1986925798425414e-10  beta= 3.9897176481869623e-13
Iteracion 21
Valores actuales: alpha= 4.205966117999996e-10  beta= 3.9897030565707805e-13
Iteracion 22
Valores actuales: alpha= 4.2111879275788673e-10  beta= 3.9896925973867083e-13
Iteracio

Iteracion 20
Valores actuales: alpha= 4.2670699248117085e-10  beta= 3.989581518832701e-13
Iteracion 21
Valores actuales: alpha= 4.254999929428202e-10  beta= 3.989605379580786e-13
Iteracion 22
Valores actuales: alpha= 4.246353454869286e-10  beta= 3.989622516747835e-13
Iteracion 23
Valores actuales: alpha= 4.2401570988419783e-10  beta= 3.989634820616433e-13
Iteracion 24
Valores actuales: alpha= 4.235715366288313e-10  beta= 3.989643652133719e-13
Iteracion 25
Valores actuales: alpha= 4.2325307749841597e-10  beta= 3.98964999011413e-13
Iteracion 26
Valores actuales: alpha= 4.2302471951325745e-10  beta= 3.9896545380101595e-13
Iteracion 27
Valores actuales: alpha= 4.2286095393622716e-10  beta= 3.98965780110784e-13
Iteracion 28
Valores actuales: alpha= 4.2274350189940894e-10  beta= 3.989660142212791e-13
Iteracion 29
Valores actuales: alpha= 4.2265926136799463e-10  beta= 3.9896618217556334e-13
Iteracion 30
Valores actuales: alpha= 4.2259883898261765e-10  beta= 3.989663026643175e-13
Iteracion 31


Iteracion 27
Valores actuales: alpha= 4.2219336572323205e-10  beta= 3.989671116936297e-13
Iteracion 28
Valores actuales: alpha= 4.222646364797619e-10  beta= 3.989669694295803e-13
Iteracion 29
Valores actuales: alpha= 4.223157651139453e-10  beta= 3.9896686738700565e-13
Iteracion 30
Valores actuales: alpha= 4.223524432413849e-10  beta= 3.989667941928155e-13
Iteracion 31
Valores actuales: alpha= 4.223787545862717e-10  beta= 3.9896674169052606e-13
Iteracion 32
Valores actuales: alpha= 4.2239762901723695e-10  beta= 3.9896670403016453e-13
Iteracion 33
Valores actuales: alpha= 4.224111684680421e-10  beta= 3.9896667701584467e-13
Iteracion 34
Valores actuales: alpha= 4.224208808470211e-10  beta= 3.9896665763797807e-13
Iteracion 35
Valores actuales: alpha= 4.22427847887274e-10  beta= 3.9896664373782473e-13
Iteracion 36
Valores actuales: alpha= 4.2243284558144533e-10  beta= 3.98966633766923e-13
Iteracion 37
Valores actuales: alpha= 4.2243643058895695e-10  beta= 3.9896662661454986e-13
Iteracion 38

### Parte 4

In [ ]:
#Usar los datos con las columnas EDA, repetir parte 3

### Parte 5

In [101]:
from sklearn.linear_model import BayesianRidge
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df_X, df_y, test_size=0.2, random_state=random_state)
BR_pipe = Pipeline([('prep',prep),('clf',BayesianRidge())])
BR_pipe.fit(X_train,y_train)
print('Fit terminado', '\n')
y_pred = BR_pipe.predict(X_test) 

R_2 = BR_pipe.score(X_test, y_test)
RMSE = np.sqrt(np.mean((y_pred-y_test)**2))
print('RMSE = ',RMSE,' R2 =',R_2,)

Fit terminado 

RMSE =  1681279.9903399977  R2 = -1.4564547221374013e-05
